### PR TITLE
fix indexes with pandas

### PR DESCRIPTION
**Pull request by imatiach-msft**
fix indexes with pandas

- add pandas indexes support, specifically for explaining timeseries models
- add datetime featurization
- fix bug in string indexing with one-hot-encoding for models that can't handle categoricals natively

### CODE REVIEW

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 1]
```diff
@@ -19,6 +21,50 @@
...
+                tmp_dataset[time_col_name + "_year"] = tmp_dataset[time_col_name].map(lambda x: x.year)
+                tmp_dataset[time_col_name + "_month"] = tmp_dataset[time_col_name].map(lambda x: x.month)
+                tmp_dataset[time_col_name + "_day"] = tmp_dataset[time_col_name].map(lambda x: x.day)
+                tmp_dataset[time_col_name + "_hour"] = tmp_dataset[time_col_name].map(lambda x: x.hour)
+                tmp_dataset[time_col_name + "_minute"] = tmp_dataset[time_col_name].map(lambda x: x.minute)
+                tmp_dataset[time_col_name + "_second"] = tmp_dataset[time_col_name].map(lambda x: x.second)
+                # Replace column itself with difference from min value, leave as same name
+                # to keep index so order of other columns remains the same for other transformations
+                tmp_dataset[time_col_name] = tmp_dataset[time_col_name].map(lambda x: x.timestamp() - self._min[idx])
+            X = tmp_dataset.values
```
Comment (wamartin-aml) [inline python/interpret_community/dataset/dataset_wrapper.py]: Do we want to return numpy if X is sparse or pandas?

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 2]
```diff
@@ -19,6 +21,50 @@
...
+
+    def transform(self, X):
+        tmp_dataset = X
+        # Temporarily convert to pandas for easier and uniform timestamp handling
+        if isinstance(X, np.ndarray):
+            tmp_dataset = pd.DataFrame(X, columns=self._features)
+        if self._time_col_names:
+            # Get the year, day, month, hour, minute, second
+            for idx, time_col_name in enumerate(self._time_col_names):
+                tmp_dataset[time_col_name + "_year"] = tmp_dataset[time_col_name].map(lambda x: x.year)
```
Comment (wamartin-aml) [inline python/interpret_community/dataset/dataset_wrapper.py]: " -> ' for all of these

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 3]
```diff
@@ -215,6 +294,25 @@ def one_hot_encode(self, columns):
...
+    def timestamp_featurizer(self):
+        """Featurizes the timestamp columns.
+
+        :return: The transformation steps to featurize the timestamp columns.
+        :rtype: DatasetWrapper
+        """
+        if self._timestamp_featurized:
+            return self._timestamp_featurizer
+        # Optimization so we don't redo this operation multiple times on the same dataset
+        self._timestamp_featurized = True
```
Comment (wamartin-aml) [inline python/interpret_community/dataset/dataset_wrapper.py]: nit: I'd prefer to set this to True just before the function returns (eliminate possible confusion when debugging this function)

File: python/interpret_community/mimic/mimic_explainer.py [hunk 4]
```diff
@@ -240,6 +252,10 @@ def __init__(self, model, initialization_examples, explainable_model, explainabl
         if not all(isinstance(categorical_feature, int) for categorical_feature in categorical_features):
             categorical_features = initialization_examples.get_column_indexes(self.features, categorical_features)
 
+        # Featurize any timestamp columns
+        # TODO: more sophisticated featurization
+        self._timestamp_featurizer = initialization_examples.timestamp_featurizer()
```
Comment (wamartin-aml) [inline python/interpret_community/mimic/mimic_explainer.py]: We're sure that this is something we want to run every time someone calls the mimic explainer?

File: python/interpret_community/mimic/mimic_explainer.py [hunk 5]
```diff
@@ -118,6 +118,8 @@ class MimicExplainer(BlackBoxExplainer):
     :type shap_values_output: interpret_community.common.constants.ShapValuesOutput
```
Comment (wamartin-aml) [inline python/interpret_community/mimic/mimic_explainer.py]: It seems like if this is something worth having for mimic_explainer it's worth having in other explainers as well?

File: test/test_mimic_explainer.py [hunk 6]
```diff
@@ -237,17 +237,18 @@ def test_explain_model_categorical(self, verify_mimic):
...
         X, _ = create_timeseries_data(sample_cnt_per_grain, 'time', 'y', grains_dict)
         model = DataFrameTestModel(X.copy())
         model = Pipeline([('test', model)])
-        features = list(X.columns)
+        features = list(X.columns.values) + list(X.index.names)
         model_task = ModelTask.Unknown
-        kwargs = {'explainable_model_args': {'n_jobs': 1}, 'augment_data': False}
+        kwargs = {'explainable_model_args': {'n_jobs': 1}, 'augment_data': False, 'reset_index': True}
+        if grains_dict:
+            kwargs['categorical_features'] = ['fruit']
```
Comment (wamartin-aml) [inline test/test_mimic_explainer.py]: Doesn't seem like nearly enough testing for the amount of custom code that was written

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 7]
```diff
@@ -19,6 +21,50 @@
...
+            for time_col_name in self._time_col_names:
+                self._min.append(tmp_dataset[time_col_name].map(lambda x: x.timestamp()).min())
+        return self
+
+    def transform(self, X):
+        tmp_dataset = X
+        # Temporarily convert to pandas for easier and uniform timestamp handling
+        if isinstance(X, np.ndarray):
+            tmp_dataset = pd.DataFrame(X, columns=self._features)
+        if self._time_col_names:
```
Comment (wamartin-aml) [inline python/interpret_community/dataset/dataset_wrapper.py]: Explicitly change to make it clear that this is a list

File: python/interpret_community/dataset/dataset_wrapper.py [hunk 8]
```diff
@@ -19,6 +21,50 @@
     from shap.common import DenseData
 
 
+class CustomTimestampFeaturizer(BaseEstimator, TransformerMixin):
+    def __init__(self, features):
+        self._features = features
+        return
+
+    def fit(self, X):
```
Comment (wamartin-aml) [inline python/interpret_community/dataset/dataset_wrapper.py]: docstrings in this class

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 2 above)
Comment (imatiach-msft) [reply python/interpret_community/dataset/dataset_wrapper.py]: done!

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 1 above)
Comment (imatiach-msft) [reply python/interpret_community/dataset/dataset_wrapper.py]: discussed offline; not dealing with sparse case, just returning in fit currently

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 3 above)
Comment (wamartin-aml) [reply python/interpret_community/dataset/dataset_wrapper.py]: nvm I see the return right below

File: python/interpret_community/mimic/mimic_explainer.py (hunk 4 above)
Comment (imatiach-msft) [reply python/interpret_community/mimic/mimic_explainer.py]: yep this was a long-standing work item, should be in other explainers eventually as well

File: python/interpret_community/mimic/mimic_explainer.py (hunk 5 above)
Comment (imatiach-msft) [reply python/interpret_community/mimic/mimic_explainer.py]: totally agree, todo for a future PR

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 7 above)
Comment (imatiach-msft) [reply python/interpret_community/dataset/dataset_wrapper.py]: done!

File: python/interpret_community/dataset/dataset_wrapper.py (hunk 8 above)
Comment (imatiach-msft) [reply python/interpret_community/dataset/dataset_wrapper.py]: done!

File: test/test_mimic_explainer.py (hunk 6 above)
Comment (imatiach-msft) [reply test/test_mimic_explainer.py]: added more tests

**Pull request merged**